sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/WitherAnnotationVisitor.java

```diff
@@ -0,0 +1,215 @@
+                    builderLocalVariable,
+                    new ExpressionDef.CallInstanceMethod(
+                        new VariableDef.This(witherType),
+                        withMethod
+                    )
+                )
+            );
+            withConsumerMethodBuilder.addStatement(
+                new ExpressionDef.CallInstanceMethod(
+                    new VariableDef.MethodParameter("consumer", consumableType),
+                    "accept", List.of(builderLocalVariable), TypeDef.VOID
+                )
+            );
+            withConsumerMethodBuilder.addStatement(
+                new StatementDef.Return(
+                    new ExpressionDef.CallInstanceMethod(
+                        builderLocalVariable,
+                        "build", List.of(), recordType
+                    )
+                )
+            );
+
+            wither.addMethod(withConsumerMethodBuilder.build());
+
+        }
+
+        SourceGenerator sourceGenerator = SourceGenerators.findByLanguage(context.getLanguage()).orElse(null);
+        if (sourceGenerator == null) {
+            return;
+        }
+
+        InterfaceDef witherDef = wither.build();
+        context.visitGeneratedSourceFile(
+            witherDef.getPackageName(),
+            witherDef.getSimpleName(),
+            recordElement
+        ).ifPresent(sourceFile -> {
+            try {
+                sourceFile.write(
+                    writer -> sourceGenerator.write(witherDef, writer)
+                );
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        });
+    }
+
+    private StatementDef withPropertyMethodStatement(ClassElement recordElement,
+                                                     ClassTypeDef recordType,
+                                                     ClassTypeDef witherType,
+                                                     TypeDef propertyTypeDef,
+                                                     PropertyElement propertyElement,
+                                                     Map<String, MethodDef> propertyAccessMethods) {
+
+        List<ExpressionDef> values = new ArrayList<>();
+        for (ParameterElement parameter : recordElement.getPrimaryConstructor().orElseThrow().getParameters()) {
+            ExpressionDef exp;
+            if (parameter.getName().equals(propertyElement.getName())) {
+                exp = new VariableDef.MethodParameter(
+                    propertyElement.getName(),
+                    propertyTypeDef
+                );
+            } else {
+                exp = new ExpressionDef.CallInstanceMethod(
+                    new VariableDef.This(witherType),
+                    propertyAccessMethods.get(parameter.getName())
+                );
+            }
+            values.add(exp);
+        }
+        return new StatementDef.Return(ExpressionDef.instantiate(recordType, values));
```

Comment:
could we add a `Objects.requireNonNull` check if the record component is annotated with `@NonNull`?

Reply:
Added for `withProperty(value)`, but not sure what to do for builders

Reply:
think we should have the same thing for builders. 

Another nice thing to support would be `@Bindable(defaultValue=..)` for builders, since currently with primitives you have to specify always the values